[17기 조혜윤]상품 View 작성

hyeyoonc opened the pull request:
>wecode PR message template 및 체크 리스트 입니다. 
아래 사항들을 전부 작성/체크 하시고 PR 해주세요!

## 수정 사항 간략한 한줄 요약
(여기에 수정 사항에 대한 간략한 한줄 요약/제목 작성해 주세요.)
- 로그인 버튼 활성화 기능 구현(예시)
		 
## 수정 사항들 자세한 내용
(여기에 수정 사항에 대한 자세한 내용을 작성해 주세요.)
- id '@' 포함, pw 5 글자 이상 입력 시 버튼 활성화(예시)

## 기타 질문 및 특이 사항
(궁금한 사항들, 하면서 느낀 점들 공유해주실 것이 있으면 자유롭게 작성해 주세요.)
- 함수 로직이 너무 긴 것 같습니다. 좀 더 효율적인 방법이 없을지 궁금합니다.(예시) 

## 체크 리스트 (아래 사항들이 전부 체크되어야만 merge가 됩니다!)
- [x] 필요한 test들을 완료하였고 기능이 제대로 실행되는지 확인 하였습니다.
- [x] Wecode의 코드 스타일 가이드에 맞추어 코드를 작성 하였습니다.
- [x] 제가 의도한 파일들과 수정 사항들만 커밋이 된 것을 확인 하였습니다.
- [x] 본 수정 사항들을 팀원들과 사전에 상의하였고 팀원들 모두 해당 PR에 대하여 알고 있습니다.

hyeyoonc committed:
"Create:ProductView 모든 상품 보기"

hyeyoonc committed:
"Modify:ProductView 최신상품보기"

hyeyoonc committed:
"Modify:ProductView 카테고리별 상품뷰"

hyeyoonc committed:
Create: GoalView healthgoal별 상품보기

hyeyoonc committed:
"Create: ProductDetailView"

hyeyoonc committed:
"Create: ProductToCartView"

hyeyoonc committed:
"Create: Product urls.py"

hyeyoonc committed:
"Modify: coreof/urls.py Add product path"

hyeyoonc committed:
"Modify: product models.py 변수명 수정"

jamessoun93 commented:
항상 PR 올리실때 하나하나 확인하고 체크해주세요 혜윤님!
<img width="574" alt="Screen Shot 2021-02-24 at 10 33 46 AM" src="https://user-images.githubusercontent.com/50141312/108931885-c7d3a480-768b-11eb-8237-90646e9bef0c.png">

jamessoun93 commented on product/views.py:
```diff
@@ -0,0 +1,277 @@
+from django.views import View
+
+from user.utils     import login_decorator
+from product.models import (Menu,
+                            Category, 
+                            Product,
+                            Image,
+                            ProductStock,
+                            RelatedProduct,
+                            Disease,
+                            ProductDisease,
+                            Allergy,
+                            ProductAllergy,
+                            Goal,
+                            ProductGoal,
+                            ProductDietaryHabit,
+                            DietaryHabit,
+                            VeganLevel,
+                            ActivityLevel,
+                            GenderCode,
+                            AgeLevel)
```
이렇게 너무 길어질 경우 아래와 같이 정렬해주실 수 있겠네요~!
```suggestion
from product.models import (
        Menu, Category, Product,
        Image, ProductStock, RelatedProduct,
        Disease, ProductDisease, Allergy,
        ProductAllergy, Goal, ProductGoal,
        ProductDietaryHabit, DietaryHabit, VeganLevel,
        ActivityLevel, GenderCode, AgeLevel
)
```

hyeyoonc replied:
수정했습니다!

jamessoun93 commented on product/urls.py:
```diff
@@ -1 +1,10 @@
+
+urlpatterns = [
+    path('', ProductView.as_view()),
+    path('/category/<int:id>', ProductView.as_view()),
```
이 부분 중간발표 때 언급드리려다가 지나가게 되었는데, 이렇게 두 가지 경우의 수가 있을 경우 path params 보다 query params 방식을 이용하여 path 를 한개만 묶어주시는게 좋습니다. 이렇게 하면 아래 view 내용과 같이 들어올지 안들어올지 모르는 id 를 None 으로 default 처리를 해줘야하는 불편함이 있고, 그 외에도 로직 상 조건에 따라 불필요하게 많은 코드가 중복되게 됩니다. 말씀드린대로 query parameter 방식으로 수정해주세요! 😉

hyeyoonc replied:
수정했습니다!

jamessoun93 commented on product/views.py:
```diff
@@ -0,0 +1,277 @@
+from order.models   import Promotion, Order, Review, OrderProductStock, OrderStatus, ShipmentStatus
+
+class ProductView(View):
+    def get(self, request, id=None):
```
urls.py 리뷰 내용 확인하신 후 query parameter 이용하는 방식으로 수정해주신다면 해당 view 에 변경사항이 많이 생길 것으로 예상되어 이 부분은 다 고치신 후 추가 리뷰드리겠습니다!

hyeyoonc replied:
네! id=None 삭제했습니다 ㅎㅎ 
삭제하고 쿼리파라미터로 모두 수정했습니다.

jamessoun93 commented on product/views.py:
```diff
@@ -0,0 +1,277 @@
+from user.models    import User, Address
+from order.models   import Promotion, Order, Review, OrderProductStock, OrderStatus, ShipmentStatus
+
+class ProductView(View):
```
저는 해당 view 내용과 아래 ProductDetailView 클래스가 있는것을 확인해서 이 view 가 상품 리스트를 전달하는 기능을 하는것을 알고있지만, class 명은 해당 class 명만 보고도 정확하게 이 class 가 어떤 기능을 담당하는 view class 인지 알 수 있게 지어주셔야 합니다. 수정해주세요! 😁

hyeyoonc replied:
ProductListView로 변경했습니다!

jamessoun93 commented on product/views.py:
```diff
@@ -0,0 +1,277 @@
+                
+                result.append(product_info)
+
+            return JsonResponse({"data": result, "message": "SUCCESS"}, status=200)
```
- `get()` 메소드를 사용하면 꼭 잡아줘야하는 exception 이 있습니다. Exception 핸들링 로직 추가해주세요!
- 지금 짜놓으신 로직으로 볼때, `request` 에서 들어온 `id` 로 매칭되는 `Goal` 객체가 존재한다면 찾아서 가공해서 데이터를 return 해주고, 아니라면 아무것도 안하고 있네요? 🧐
- 이런 부분 해결된 후 디테일하게 로직 살펴보고 리뷰할게요~!

jamessoun93 commented on product/views.py:
```diff
@@ -0,0 +1,277 @@
+    def get(self, request, product_id):
+        if not Product.objects.filter(id=product_id).exists():
+            return JsonResponse({'message': 'DOES_NOT_EXIST'}, status=404)
+        product  = Product.objects.get(id=product_id)
```
마찬가지로 exception handling 로직 추가해주세요!

jamessoun93 commented on product/views.py:
```diff
@@ -0,0 +1,277 @@
+        category = product.category.menu.name
+
+        vegan_level = product.vegan_level_id
+        if vegan_level == 1:
+            is_vegan      = True
+            is_vegeterian = False
+        elif vegan_level == 2:
+            is_vegan      = False
+            is_vegeterian = True
+        else:
+            is_vegan      = False
+            is_vegeterian = False
```
이 부분은 조금 정리가 필요할 것 같습니다. else 에 해당하는 옵션은 뭐가 있을까요? `3` 이상인가요 아니면 아예 `NULL` 일까요?

yangahh replied:
값이 1, 2, 3 뿐이라서 3인경우는 else로 처리하였습니다.

jamessoun93 commented on product/views.py:
```diff
@@ -0,0 +1,277 @@
+        else:
+            price = {
+                product_SSPs[0].size : product_SSPs[0].price,
+                product_SSPs[1].size : product_SSPs[1].price
+            }
+            is_soldout = {
+                product_SSPs[0].size : bool(product_SSPs[0].stock == 0),
+                product_SSPs[1].size : bool(product_SSPs[1].stock == 0)
+            }
+
+        similar_product_list = []
+        goals_product        = product.goal.all()
+        for goal_product in goals_product:
+            similar_products = goal_product.product_set.exclude(id=product_id)
+            for similar_product in similar_products:
+                similar_product_info = {
+                        'id'             : similar_product.id,
+                        'title'          : similar_product.name,
+                        'subTitle'       : similar_product.sub_name,
+                        'imageUrl'       : similar_product.image_set.get(is_main=True).image_url,
+                        'healthGoalList' : [goal.name for goal in similar_product.goal.all()]
+                }
+                similar_product_list.append(similar_product_info)
+        context = {}
+        context['category']            = category 
+        context['productImageSrc']     = product.image_set.get(is_main=False).image_url
+        context['productCardImageSrc'] = product.image_set.get(is_main=True).image_url
+        context['isVegan']             = is_vegan
+        context['isVegetarian']        = is_vegeterian
+        context['healthGoalList']      = [goal.name for goal in product.goal.all()]
+        context['title']               = product.name
+        context['subTitle']            = product.sub_name
+        context['description']         = product.description
+        context['nutritionLink']       = product.nutrition_url
+        context['allergyList']         = [allergy.name for allergy in product.allergy.all()]
+        context['productPrice']        = price
+        context['isSoldOut']           = is_soldout
+        context['dietaryHabitList']    = [dietary_habit.name for dietary_habit in product.dietary_habit.all()]
+        context['similarProduct']      = similar_product_list[:2] 
+        return JsonResponse({"data": context, "message": "SUCCESS"}, status=200)
```
이 부분 전체적으로 flow 를 직접 혜윤님께 듣고 정리할 수 있는 방법을 같이 찾아봐야할 것 같습니다. 이따 호출할게요!

jamessoun93 commented on product/views.py:
```diff
@@ -0,0 +1,277 @@
+    def post(self, request):
+            data          = json.loads(request.body)
+            user          = request.user
+            product_id    = data.get('productId', None)
+            product_size  = data.get('productSize', None)
+            product_price = data.get('productPrice', None)
+            if not (product_id and product_price): 
+                return JsonResponse({"message": "KEY_ERROR"}, status=400)            
```
`get()` 을 쓰는 이유가 없어도 되는 값을 지정하신 default 값으로 처리해주기 위함인데, 지금 로직 상 `product_id` 와 `product_price` 는 필수값 아닌가요?!

hyeyoonc replied:
KeyError 처리한건데 안될까요 ㅠㅜㅎㅎㅎㅎ

jamessoun93 commented on product/views.py:
```diff
@@ -0,0 +1,277 @@
+            product_price = data.get('productPrice', None)
+            if not (product_id and product_price): 
+                return JsonResponse({"message": "KEY_ERROR"}, status=400)            
+            if not ProductStock.objects.filter(product_id=product_id, size=product_size):
```
해당 QuerySet 리스트에 ProductStock 객체의 존재유무를 판단하는 용도이면 filter 에서 끝나는 것이 아니라 `exists()` 메소드를 추가로 활용해주시면 값을 전부 가져오는것이 아니라 존재유무만 확인할 수 있게됩니다!

jamessoun93 commented on product/views.py:
```diff
@@ -0,0 +1,277 @@
+            if not ProductStock.objects.filter(product_id=product_id, size=product_size):
+                return JsonResponse({"message": "PRODUCT_DOES_NOT_EXIST"}, status=400)
+            product_price = float(product_price)
+            if not Order.objects.filter(user=user, order_status_id=1).exists():
+                shipping_cost = 5 if product_price < 20 else 0
+                order_info    = Order.objects.create( 
+                    user            = user,
+                    order_number    = datetime.today().strftime("%Y%m%d") + str(randint(100000, 1000000)),
+                    order_status_id = 1,
+                    sub_total_cost  = product_price,
+                    shipping_cost   = shipping_cost,
+                    total_cost      = product_price + shipping_cost
+                )
+            else:
+                order_info = Order.objects.get(user=user, order_status_id=1)
+                order_info.sub_total_cost = float(order_info.sub_total_cost) + product_price
+                order_info.shipping_cost  = 5 if order_info.sub_total_cost < 20 else 0
+                order_info.total_cost     = float(order_info.sub_total_cost) + order_info.shipping_cost
+                order_info.save()
```
아래에서는 `update_or_create()` 활용하셨는데, 여기서는 안하시는 이유가 있나요?!

yangahh replied:
update 할 때는 sub_total_cost와 total_cost가 기존 가격에 새로 받은 가격이 더해져야하는데(기존 가격 : order_info.sub_total_cost)
update_or_create로 처리하면 create 할 때는  order_info.sub_total_cost로 처리할 수가 없어서 로직을 나눴습니다.

다른 방법이 있으면 알려주시면 정말 감사할것같아요ㅜㅜㅎㅎ

jamessoun93 commented:
혜윤님 정말 잘 작성해주셨네요! 다만 파이썬의 가장 큰 장점인 가독성을 좋게 유지하기 위해 빈 줄을 꼭 추가해주세요!

hyeyoonc committed:
"Modify: product models.py core/of/urls.py"

hyeyoonc committed:
Modify:product/views.py urls.py

hyeyoonc committed:
Modify: urls.py에서 불필요한 import 삭제

hyeyoonc committed:
Modify: product/views.py
불필요한 import 모듈 수정
쿼리 파라미터 적용
class명 ProductView에서 ProductListView로 수정
ProductDetailView flow 수정

hyeyoonc committed:
Modify:login_decorator 메세지 수정

hyeyoonc committed:
Modify: product/url.py 클래스 이름변경으로 url변경

hyeyoonc committed:
Modify:ProductView 오타수정"

hyeyoonc committed:
Modify: product/view의 ProductListView 수정

jamessoun93 commented on product/views.py:
```diff
@@ -0,0 +1,208 @@
+        sort_dic["all"]            = ""
+
+        result = []
+        categories = [categories for categories in Category.objects.filter(name__contains=sort_dic[sort])]
```
구두로 전달.

jamessoun93 commented on product/views.py:
```diff
@@ -0,0 +1,208 @@
+                products = Product.objects.filter(category=category)  
+                product_info_list = self.make_product_info_list(products=products) 
+                context["item"]   = product_info_list
+                result.append(context)
```
list comprehension 으로 변경!

hyeyoonc committed:
Modify: ProductListView 코드 리팩토링

hyeyoonc committed:
Modify: response 데이터 추가

Kang-kyunghun commented on product/urls.py:
```diff
@@ -1 +1,8 @@
+urlpatterns = [
+    path('', ProductListView.as_view()),
+    path('/<int:product_id>', ProductDetailView.as_view()),
+    path('/tocart', ProductToCartView.as_view()),
```
order app으로 옯겨 주세요.

Kang-kyunghun commented on product/views.py:
```diff
@@ -0,0 +1,193 @@
+            product_size  = data.get('productSize', None)
+            product_price = data.get('productPrice', None)
+
+            if not (product_id and product_price): 
```
key error try except로 바꿔 주세요.

Kang-kyunghun commented on product/views.py:
```diff
@@ -0,0 +1,193 @@
+
+            # orders insert
+            if not Order.objects.filter(user=user, order_status_id=1).exists():
+                shipping_cost = 5 if product_price < 20 else 0
```
상수 처리 해주세요~~!!!

Kang-kyunghun commented on product/views.py:
```diff
@@ -0,0 +1,193 @@
+                order         = Order.objects.create(  # order_info
+                    user            = user,
+                    order_number    = datetime.today().strftime("%Y%m%d") + str(randint(1000, 10000)),
+                    order_status_id = 1,
```
```suggestion
                    order_status = OrderStatus.objects.get({조건}),
```

Kang-kyunghun commented on product/views.py:
```diff
@@ -0,0 +1,193 @@
+            q.add(Q(name__icontains=sort),q.OR)
+        categories = Category.objects.filter(q)
+
+        if categories: 
```
```suggestion
        if categories.exsits(): 
```

Kang-kyunghun commented on product/views.py:
```diff
@@ -0,0 +1,193 @@
+
+        if Goal.objects.filter(name__icontains=sort).exists():
+            goal     = Goal.objects.get(name__icontains=sort)
+            products = Product.objects.filter(goal=goal)
```
goal로 product 가져오기

Kang-kyunghun commented on product/views.py:
```diff
@@ -0,0 +1,193 @@
+                "isNew"        : product.is_new,
+                "isSoldout"    : is_soldout
+            }
+            product_info_list.append(product_info)
```
list comprehension으로 바꿔 주세요.

Kang-kyunghun commented on product/views.py:
```diff
@@ -0,0 +1,193 @@
+
+        context  = {}
+        product  = Product.objects.get(id=product_id)
+        category = product.category.menu.name
```
category_name은 어떨까요?

hyeyoonc replied:
좋아요

Kang-kyunghun commented on product/views.py:
```diff
@@ -0,0 +1,193 @@
+        # size, price
+        product_SSPs = ProductStock.objects.filter(product=product)  # SSP: siz e, stock, price
+        if category == 'vitamins':
+            price      = product_SSPs[0].price   
+            is_soldout = bool(product_SSPs[0].stock == 0)
```
```suggestion
            price      = product_SSPs.first().price   
            is_soldout = bool(product_SSPs.first().stock == 0)
```

Kang-kyunghun commented on product/views.py:
```diff
@@ -0,0 +1,193 @@
+            price      = product_SSPs[0].price   
+            is_soldout = bool(product_SSPs[0].stock == 0)
+        else:
+            price = {
+                product_SSPs[0].size : product_SSPs[0].price,
+                product_SSPs[1].size : product_SSPs[1].price
+            }
+            is_soldout = {
+                product_SSPs[0].size : bool(product_SSPs[0].stock == 0),
+                product_SSPs[1].size : bool(product_SSPs[1].stock == 0)
+            }
```
```suggestion
            price = {product_SSP.size : product_SSP.price for product_SSP in product_SSPs}
            is_soldout = {product_SSP.size : bool(product_SSP.stock == 0) for product_SSP in product_SSPs}

```

Kang-kyunghun commented on product/views.py:
```diff
@@ -0,0 +1,193 @@
+
+        context['productPrice']     = price
+        context['isSoldOut']        = is_soldout
+        context['dietaryHabitList'] = [dietary_habit.name for dietary_habit in product.dietary_habit.all()]
```
context ={ key1 : value1, key2 : value2} 이런 식으로 표현 하시고,
로직이 관여된 부분은 context['key_logic']=value_logic

Kang-kyunghun commented on product/views.py:
```diff
@@ -0,0 +1,193 @@
+        context['dietaryHabitList'] = [dietary_habit.name for dietary_habit in product.dietary_habit.all()]
+
+        # similar products
+        similar_product_list = []
+        goals_product        = product.goal.all()
+        for goal_product in goals_product:
+            similar_products = goal_product.product_set.exclude(id=product_id)
+
+            for similar_product in similar_products:
+                similar_product_info = {
+                        'id'             : similar_product.id,
+                        'title'          : similar_product.name,
+                        'subTitle'       : similar_product.sub_name,
+                        'imageUrl'       : similar_product.image_set.get(is_main=True).image_url,
+                        'healthGoalList' : [goal.name for goal in similar_product.goal.all()]
+                }
+                similar_product_list.append(similar_product_info)
```
list comprehension으로 바꿔 주세요.

Kang-kyunghun replied:
```suggestion
        for goal_product in goals_products:
            similar_product_list = [                                                                                                                                                                      
          {                                                                                                                                                                                     
              'id': similar_product.id,                                                                                                                                                         
               'title': similar_product.name,                                                                                                                                                    
              'subtitle': similar_product.sub_name,                                                                                                                                             
              'imageUrl': similar_product.image_set.get(is_main=True).image_url,                                                                                                                
              'healthGoalList': [goal.name for goal in similar_product.goal.all()]                                                                                                              
           } for similar_product in goal_product.product_set.exclude(id=product_id)]                                                                                                           
```
밖에 for 문만 list comprehension으로 바꾸고 안에 for문은 살리는 방향으로 가면 될 것같아요
둘 다 하는 방법이 있기는 한데 그건 가독성이 너무 떨어져서 속도만 조금 개선 시키는 방향으로 가면 될 것 같습니다.

hyeyoonc committed:
Modify:list comprehension으로 변경

Kang-kyunghun commented on product/views.py:
```diff
@@ -0,0 +1,182 @@
+
+        return JsonResponse({"data": context, "message": "SUCCESS"}, status=200)
+
+#Add 눌렀을 때 
+class ProductToCartView(View):
+    @login_decorator
+    def post(self, request):
```
해당 기능은 장바구니와 관련된 뷰에 post method에 더 적합 합니다. 
옮겨 주세요.

hyeyoonc replied:
네 ㅎㅎ

Kang-kyunghun commented on product/views.py:
```diff
@@ -0,0 +1,182 @@
+
+            # orders insert
+            if not Order.objects.filter(user=user, order_status=OrderStatus.objects.get(name='주문 전')).exists():
+                shipping_cost = SHIPPING_COST if product_price < 20 else 0
```
SHIPPING_COST만 상수처리 하였는데, 20, 0도 상수처리 해주세요.

hyeyoonc replied:
네~

Kang-kyunghun commented on product/views.py:
```diff
@@ -0,0 +1,182 @@
+            else:
+                order = Order.objects.get(user=user, order_status=OrderStatus.objects.get(name='주문 전'))
+                order.sub_total_cost = float(order.sub_total_cost) + product_priceㅑ
+                order.shipping_cost  = SHIPPING_COST if order.sub_total_cost < 20 else 0
```
SHIPPING_COST만 상수처리 하였는데, 20, 0도 상수처리 해주세요.

hyeyoonc replied:
네~

Kang-kyunghun commented on product/views.py:
```diff
@@ -0,0 +1,182 @@
+                order.total_cost     = float(order.sub_total_cost) + order.shipping_cost
+                order.save()
+            
+            if not ProductStock.objects.filter(product_id=product_id, size=product_size):
```
```suggestion
            if not ProductStock.objects.filter(product_id=product_id, size=product_size).exist():
```
조건에 대한 판단을 위해 objects.filter을 사용할 때 .exists()을 붙여줘야 모든 값을 다 가져오지 않고, 존재 여부만 판단 합니다.

hyeyoonc replied:
네~~

Kang-kyunghun commented on product/views.py:
```diff
@@ -0,0 +1,182 @@
+                return JsonResponse({"message": "PRODUCT_DOES_NOT_EXIST"}, status=400)
+
+            # order_product_stocks insert
+            added_product = ProductStock.objects.get(product_id=product_id, size=product_size)
```
objects.get()을 사용할 때 발생할 수 있는 에러에 대해 예외 처리 추가해 주세요.

hyeyoonc replied:
처리했습니다

Kang-kyunghun commented on product/views.py:
```diff
@@ -0,0 +1,182 @@
+        context['isSoldOut']        = is_soldout
+
+        # similar products
+        goals_product        = product.goal.all()
+        for goal_product in goals_product:
```
```suggestion
        for goal_product in product.goal.all():
```
goals_product 해당 변수가 여러번 사용 되지 않기 때문에 변수를 선언하지 않아도 될 것으로 보입니다.

hyeyoonc committed:
Modify: product views.py ProductToCartView 삭제

hyeyoonc committed:
Modify: ProductToCartView url 삭제

hyeyoonc committed:
Modify: 오타 수정

hyeyoonc committed:
Modify: try except 처리

soheon-lee commented on product/views.py:
```diff
@@ -0,0 +1,131 @@
+            product_SSPs = ProductStock.objects.filter(product=product)  # SSP: siz e, stock, price
+            if category_name == 'vitamins':
+                price      = product_SSPs.first().price   
+                is_soldout = bool(product_SSPs.first().stock == 0)
```
```suggestion
                is_soldout = product_SSPs.first().stock == 0
```
연산자의 return은 이미 불리언 타입이기 때문에 따로 형변환을 하는 과정을 거치는 것은 같은 작업을 두번 수행하는 것입니다.
하지 않으셔도 됩니다 :)

soheon-lee commented on product/views.py:
```diff
@@ -0,0 +1,131 @@
+                is_soldout = bool(product_SSPs.first().stock == 0)
+            else:
+                price      = {product_SSP.size : product_SSP.price for product_SSP in product_SSPs}
+                is_soldout = {product_SSP.size : bool(product_SSP.stock == 0) for product_SSP in product_SSPs}
```
```suggestion
                is_soldout = {product_SSP.size : product_SSP.stock == 0 for product_SSP in product_SSPs}
```
여기도 마찬가지입니다

hyeyoonc replied:
네~

soheon-lee commented on product/views.py:
```diff
@@ -0,0 +1,131 @@
+
+        # similar products
+            for goal_product in product.goal.all():
+                similar_product_list = [
+                    {
+                        'id'            : similar_product.id,
+                        'title'         : similar_product.name,
+                        'subTitle'      : similar_product.sub_name,
+                        'imageUrl'      : similar_product.image_set.get(is_main=True).image_url,
+                        'healthGoalList': [goal.name for goal in similar_product.goal.all()],
+
+                        }
```
```suggestion
                similar_product_list = [
                    {
                        'id'            : similar_product.id,
                        'title'         : similar_product.name,
                        'subTitle'      : similar_product.sub_name,
                        'imageUrl'      : similar_product.image_set.get(is_main=True).image_url,
                        'healthGoalList': [goal.name for goal in similar_product.goal.all()]
                    }
```
필요없는 공백은 제거해주시고, 같은 레벨의 데이터를 감싸고 있는 괄호는 위치를 맞춰주시는 것이 좋습니다 👍

hyeyoonc replied:
네 수정하겠습니다

soheon-lee commented on product/views.py:
```diff
@@ -0,0 +1,131 @@
+                        'healthGoalList': [goal.name for goal in similar_product.goal.all()],
+
+                        }
+                    for similar_product in goal_product.product_set.exclude(id=product_id)]
```
`exclude` good!

soheon-lee commented on product/urls.py:
```diff
@@ -1 +1,7 @@
+from product.views import ProductListView, ProductDetailView
+
+urlpatterns = [
+    path('', ProductListView.as_view()),
+    path('/<int:product_id>', ProductDetailView.as_view()),
```
RESTful URL good good! 👍

hyeyoonc committed:
Modify:bool 타입 변경

hyeyoonc committed:
Modify:bool 타입 수정

hyeyoonc committed:
Modify:is_new 조건 수정

soheon-lee commented:
LGTM!
